Mono.Debugging.Soft/SoftDebuggerSession.cs

```diff
@@ -2640,58 +2640,67 @@ bool LoadPdbFile (string pdbFileName)
-						list = new List<SequencePoint []> ();
-
-					SourceFileDebugInfo info = new SourceFileDebugInfo (list);
-
-					var document = pdb.GetDocument (documentHandle);
-					info.Hash = pdb.GetBlobBytes (document.Hash);
-					info.HashAlgorithem = pdb.GetGuid (document.HashAlgorithm);
-					info.FileID = 0;
-					info.FullFilePath = pdb.GetString (document.Name);
-
-					fileToSourceFileInfos [info.FullFilePath] = new List<SourceFileDebugInfo> ();
-
-					if (!fileToSourceFileInfos.ContainsKey (Path.GetFileName (info.FullFilePath)))
-						fileToSourceFileInfos [Path.GetFileName (info.FullFilePath)] = new List<SourceFileDebugInfo> ();
-
-					fileToSourceFileInfos [info.FullFilePath].Add (info);
-					fileToSourceFileInfos [Path.GetFileName (info.FullFilePath)].Add (info);
-				}
-			}
-
-			return true;
-		}
-
+			var fileToSourceFileInfos = new Dictionary<string, List<SourceFileDebugInfo>>(StringComparer.InvariantCultureIgnoreCase);
+			sourceFilesDebugInfo[pdbFileName] = fileToSourceFileInfos;
+
+
+			using (var module = Cecil.ModuleDefinition.ReadModule(Path.ChangeExtension(pdbFileName, ".dll")))
+			{
+				var symbolReaderProvider = new Cecil.Cil.DefaultSymbolReaderProvider(false);
+				module.ReadSymbols(symbolReaderProvider.GetSymbolReader(module, pdbFileName));
+
+				var methodMapping = new Dictionary<Document, List<SequencePoint[]>>();
+				foreach (var type in module.Types)
+				{
+					LoadPdbType(type, fileToSourceFileInfos);
+
+					foreach (var nestedType in type.NestedTypes)
+						LoadPdbType(nestedType, fileToSourceFileInfos);
+				}
+			}
+
+			return true;
+		}
+
+		private void LoadPdbType(Cecil.TypeDefinition type, Dictionary<string, List<SourceFileDebugInfo>> fileToSourceFileInfos)
+		{
+			foreach (var method in type.Methods)
+			{
+				var documents = method.DebugInformation.SequencePoints.GroupBy(t => t.Document, t => t, (document, sequencePoints) => new { Document = document, SequencePoints = sequencePoints });
+
+				foreach (var doc in documents)
+				{
+					AddSequencePointsToFileSource(fileToSourceFileInfos, doc.Document, doc.SequencePoints, doc.Document.Url);
+					AddSequencePointsToFileSource(fileToSourceFileInfos, doc.Document, doc.SequencePoints, Path.GetFileName(doc.Document.Url));
+				}
+			}
+		}
+
+		private static void AddSequencePointsToFileSource(Dictionary<string, List<SourceFileDebugInfo>> fileToSourceFileInfos, Document document, IEnumerable<SequencePoint> sequencePoints, string name)
+		{
+			if (!fileToSourceFileInfos.ContainsKey(name))
+			{
+				List<SequencePoint[]> list = new List<SequencePoint[]>(new[] { sequencePoints.ToArray() });
+				SourceFileDebugInfo info = new SourceFileDebugInfo(list);
+
+				info.Hash = document.Hash;
+				info.FileID = 0;
+				info.FullFilePath = document.Url;
+
+				fileToSourceFileInfos.Add(name, new List<SourceFileDebugInfo> { info });
+			}
+			else
+			{
+				fileToSourceFileInfos[name].First().PdbMethods.Add(sequencePoints.ToArray());

```

Comment:
Even though I agree on the general statement I didn't want to change existing code. The hileToSource is a dictionary of list of array, where the list must have only one item (there's a check in CheckBestMatch to see if the list has more than one item. As before, even though I think that's really not ideal I didn't want to change existing code too much. I can do it if it's preferred.